File: backend/app/schemas/meeting.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


MeetingStatus = Literal["queued", "running", "completed", "failed"]


class ApiSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ParticipantCreate(ApiSchema):
    name: str = Field(max_length=255)

    @field_validator("name")
    @classmethod
    def strip_and_validate_name(cls, value: str) -> str:
        name = value.strip()
        if not name:
            raise ValueError("Participant name must not be empty")
        return name


class MeetingCreateResponse(ApiSchema):
    id: UUID
    status: MeetingStatus


class MeetingListItem(ApiSchema):
    id: UUID
    title: str
    started_at: datetime
    status: MeetingStatus
    confirmed_at: datetime | None


class MeetingListResponse(ApiSchema):
    items: list[MeetingListItem]


class ParticipantResponse(ApiSchema):
    id: UUID
    name: str


class SpeakerResponse(ApiSchema):
    label: str
    participant_id: UUID | None


class SegmentResponse(ApiSchema):
    id: UUID
    speaker: str | None
    start_ms: int
    end_ms: int
    text: str


class ActionItemResponse(ApiSchema):
    id: UUID
    text: str
    assignee_id: UUID | None
    due_date: date | None
    needs_review: bool
    review_reason: str | None
    source_segment_ids: list[UUID]


class MeetingDetailResponse(ApiSchema):
    id: UUID
    title: str
    started_at: datetime
    timezone: str
    status: MeetingStatus
    error: str | None
    confirmed_at: datetime | None
    summary: str | None
    participants: list[ParticipantResponse]
    speakers: list[SpeakerResponse]
    segments: list[SegmentResponse]
    action_items: list[ActionItemResponse]

File: frontend/src/types/meeting.ts
export type MeetingStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed";

export interface Participant {
  id: string;
  name: string;
}

export interface SpeakerMapping {
  label: string;
  participant_id: string | null;
}

export interface TranscriptSegment {
  id: string;
  speaker: string | null;
  start_ms: number;
  end_ms: number;
  text: string;
}

export interface ActionItem {
  id: string;
  text: string;
  assignee_id: string | null;
  due_date: string | null;
  needs_review: boolean;
  review_reason: string | null;
  source_segment_ids: string[];
}

export interface Meeting {
  id: string;
  title: string;
  started_at: string;
  timezone: string;
  status: MeetingStatus;
  error: string | null;
  confirmed_at: string | null;
  summary: string | null;
  participants: Participant[];
  speakers: SpeakerMapping[];
  segments: TranscriptSegment[];
  action_items: ActionItem[];
  audio_url?: string | null;
}

export interface MeetingListItem {
  id: string;
  title: string;
  started_at: string;
  status: MeetingStatus;
  confirmed_at: string | null;
}

export interface MeetingListResponse {
  items: MeetingListItem[];
}

export interface CreateMeetingResponse {
  id: string;
  status: MeetingStatus;
}
